internal/api/mfa.go

```diff
@@ -61,6 +85,168 @@ const (
+	}
+
+	return webAuthn, nil
+}
+
+func (a *API) handleWebAuthnEnrollment(w http.ResponseWriter, r *http.Request, params *EnrollFactorParams) error {
+	ctx := r.Context()
+	user := getUser(ctx)
+	db := a.db.WithContext(ctx)
+	ipAddress := utilities.GetIPAddress(r)
+
+	webAuthn, err := initializeWebAuthnConfig(params.WebAuthn)
+	if err != nil {
+		return err
+	}
+	options, session, err := webAuthn.BeginRegistration(user)
+	if err != nil {
+		return internalServerError("error generating webauthn registration data").WithInternalError(err)
+	}
+	ws := &models.WebauthnSession{
+		SessionData: session,
+	}
+	factor := models.NewFactor(user, params.FriendlyName, models.WebAuthn, models.FactorStateUnverified)
+	challenge := ws.ToChallenge(factor.ID, ipAddress)
+	err = db.Transaction(func(tx *storage.Connection) error {
+		if terr := tx.Create(factor); err != nil {
+			return terr
+		}
+		if terr := tx.Create(challenge); terr != nil {
+			return terr
+		}
+		return nil
+
+	})
+	if err != nil {
+		return err
+	}
+
+	return sendJSON(w, http.StatusOK, &EnrollWebAuthnFactorResponse{
+		PublicKeyCredentialRequestOptions: options,
+		FactorID:                          factor.ID,
+		ChallengeID:                       challenge.ID,
+		FriendlyName:                      factor.FriendlyName,
+	})
+
+}
+
+func (a *API) handleWebAuthnVerification(w http.ResponseWriter, r *http.Request, params *VerifyFactorParams) error {
+	ctx := r.Context()
+	user := getUser(ctx)
+	config := a.config
+	factor := getFactor(ctx)
+	webAuthn, err := initializeWebAuthnConfig(&params.WebAuthn)
+	if err != nil {
+		return err
+	}
+	db := a.db.WithContext(ctx)
+	challenge, err := models.FindChallengeByID(a.db, params.ChallengeID)
+	if err != nil {
+		return err
+	}
+	sessionData := challenge.ToSession(user.ID, config.MFA.ChallengeExpiryDuration)
+
+	var credential *webauthn.Credential
+
+	if factor.IsUnverified() {
+		credential, err = webAuthn.FinishRegistration(user, sessionData, r)
+		if err != nil {
+			return err
+		}
+
+	} else if factor.IsVerified() {
+		credential, err = webAuthn.FinishLogin(user, sessionData, r)
+		if err != nil {
+			return internalServerError("error validating webauthn credentials")
+		}
+	}
+
+	var token *AccessTokenResponse
+	err = db.Transaction(func(tx *storage.Connection) error {
+		var terr error
+		if terr = models.NewAuditLogEntry(r, tx, user, models.VerifyFactorAction, r.RemoteAddr, map[string]interface{}{
+			"factor_id":    factor.ID,
+			"challenge_id": challenge.ID,
+			"factor_type":  factor.FactorType,
+		}); terr != nil {
+			return terr
+		}
+		if terr = challenge.Verify(tx); terr != nil {
+			return terr
+		}
+		if !factor.IsVerified() {
+			if terr = factor.UpdateStatus(tx, models.FactorStateVerified); terr != nil {
+				return terr
+			}
+			if terr = factor.UpdateWebauthnFactorPostVerification(tx, credential.PublicKey, credential.ID); terr != nil {
```

Comment:
As far as I can see `credential` is a JSON-encodable struct. So maybe we can just put the whole credential in the database as `jsonb`?